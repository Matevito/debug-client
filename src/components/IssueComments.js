import React from 'react'

// mui comp
import {
    Box,
    Card,
    CardContent,
    TableContainer,
    Table,
    TableHead,
    TableCell,
    TableBody,
    TableRow,
    Typography,
    Paper,
    Link
} from "@mui/material"

// app components
import { detailedDate } from "../features/dateFormatter";
import { CommentForm } from "./CommentForm";

const Row = ({ comment }) => {
    return (
        <TableRow sx={{ '& > *': { borderBottom: 'unset'}, }} >
            <TableCell component="th" scope="row">
                <Typography>{comment.user.username}</Typography>
            </TableCell>
            <TableCell align="left" colSpan={4}>
                <Typography><div dangerouslySetInnerHTML={{__html: comment.message}} /></Typography>
            </TableCell>
            <TableCell align="center">
                <Typography>
                    {comment.screenshots.length ? 
                        comment.screenshots.map((url, index) => {
                            return(
                                <Link
                                    key={index}
                                    href={url}
                                    target="_blank"
                                    rel="noreferrer"
                                    style={{ color: "inherit"}} underline="hover"
                                >
                                    {index+1}
                                </Link>
                            )
                        })
                        :"None"
                    }
                </Typography>
            </TableCell>
            <TableCell align="left">
                <Typography>{detailedDate(comment.date)}</Typography>
            </TableCell>
        </TableRow>
    )
}

export const IssueComments = ({ comments, handleSubmit }) => {
    return (
        <Box sx={{marginTop: "30px" }}>
            
            <TableContainer component={Paper} raised="true"
                sx={{ 
                    height: 455,
                    backgroundColor:"#c4c1c1",
                    Width: "100%"
                }}
            >
                <Table>
                    <TableHead>
                        <TableRow>
                            <TableCell align="center" colSpan={7}>
                                <Typography><b>Issue Comments</b></Typography>
                            </TableCell>
                        </TableRow>
                        <TableRow>
                            <TableCell align="left">
                                <Typography><b>User</b></Typography>
                            </TableCell>
                            <TableCell align="left" colSpan={4}> 
                                <Typography><b>Comment</b></Typography>
                            </TableCell>
                            <TableCell align="left">
                                <Typography><b>Images</b></Typography>
                            </TableCell>
                            <TableCell align="left">
                                <Typography><b>Created</b></Typography>
                            </TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {comments.slice(0).reverse().map((comment, index) => {
                            return (
                                <Row key={index} comment={comment} />
                            )
                        })}
                    </TableBody>
                </Table>
            </TableContainer>

            <Card raised={true} sx={{marginTop: "30px"}}>
                <CardContent sx={{backgroundColor: "#dbd7cc",}}>
                    <CommentForm handleSubmit={handleSubmit}/>
                </CardContent>
            </Card>
        </Box>
    )
}
