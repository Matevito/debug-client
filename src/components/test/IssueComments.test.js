import React from "react";

// testing comp
import { fireEvent, render, screen } from "@testing-library/react";
import '@testing-library/jest-dom'

// test comp
import { IssueComments } from "../IssueComments";

describe("IssueComments component", () => {
    const noComments = []
    let comments;
    beforeEach(() => {
        jest.resetAllMocks();
        comments = [
            {
                _id: "uno",
                message: "the first message",
                screenshots: [],
                user: { _id: "1", username: "testUser1"},
                date: "2022-05-02T19:37:35.307Z",
            },
            {
                _id: "does",
                message: "second one",
                screenshots: [],
                user: { _id: "2", username: "adminUser"},
                date: "2022-05-02T19:37:35.307Z",
            },
            {
                _id: "tres",
                message: "you get it...",
                screenshots: [],
                user: { _id: "1", username: "testUser1"},
                date: "2022-05-02T19:37:35.307Z",
            }
        ]
    })
    test("comp handles blank comments array", () => {
        const { container } = render(<IssueComments comments={[]} />)
        expect(container).toMatchSnapshot();
    });
    test("comp renders basic comments", () => {
        render(<IssueComments comments={comments} />)

        // renders header
        expect(screen.getByText("Issue Comments")).toBeInTheDocument()
        expect(screen.getByText("User")).toBeInTheDocument()
        expect(screen.getByText("Comment")).toBeInTheDocument()
        expect(screen.getByText("Images")).toBeInTheDocument()
        expect(screen.getByText("Created")).toBeInTheDocument()

        // renders body
        expect(screen.getByText("adminUser")).toBeInTheDocument()
        expect(screen.getByText("you get it...")).toBeInTheDocument()
        expect(screen.getByText("second one")).toBeInTheDocument()
        expect(screen.getAllByText("None").length).toBe(3)
        expect(screen.getAllByText("testUser1").length).toBe(2)
        expect(screen.getAllByText("May 2, 2022, 2:37:35 PM GMT-5").length).toBe(3)
    });
    test("comp handles comments with screenshots", () => {
        comments[0].screenshots = ["testUrl1", "testUrl2"];
        render(<IssueComments comments={comments} />)

        expect(screen.getByText("1")).toBeInTheDocument()
        expect(screen.getByText("2")).toBeInTheDocument()
    });
    test.todo("renders correct form");
});