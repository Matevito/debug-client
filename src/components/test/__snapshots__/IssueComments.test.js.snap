// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueComments component comp handles blank comments array 1`] = `
<div>
  <div
    class="MuiBox-root css-46b038"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-sxm50k-MuiPaper-root-MuiTableContainer-root"
      raised="true"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              colspan="7"
              scope="col"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                <b>
                  Issue Comments
                </b>
              </p>
            </th>
          </tr>
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                <b>
                  User
                </b>
              </p>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              colspan="4"
              scope="col"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                <b>
                  Comment
                </b>
              </p>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                <b>
                  Images
                </b>
              </p>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                <b>
                  Created
                </b>
              </p>
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        />
      </table>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiCard-root css-c8lkdx-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root css-1u8j0in-MuiCardContent-root"
      >
        <div
          class="MuiBox-root css-dvxtzn"
        />
        <div
          class="MuiBox-root css-q2ah1r"
        />
        <form
          action="#"
        >
          <div
            class="MuiBox-root css-2n86dw"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                Message
              </label>
              <div
                class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-10i04qz-MuiInputBase-root-MuiFilledInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiFilledInput-input MuiInputBase-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
                  id=":r0:"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <label
              class="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root  css-1ph36rw-MuiButtonBase-root-MuiButton-root"
              role="button"
              tabindex="0"
              type="file"
            >
              <input
                hidden=""
                type="file"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="UploadIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M5 20h14v-2H5v2zm0-10h4v6h6v-6h4l-7-7-7 7z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </label>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root  css-11qr2p8-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SendIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
